File: lib/services/websocket_service.dart
import 'dart:async';
import 'dart:convert';
import 'package:flutter/foundation.dart';
import 'package:web_socket_channel/web_socket_channel.dart';
import 'package:web_socket_channel/io.dart';
import 'package:appdev_md/models/motion_event.dart';

/// Service for handling WebSocket connections to the backend
class WebSocketService {
  /// The WebSocket channel
  WebSocketChannel? _channel;

  /// Stream controller for motion events
  final _motionEventsController = StreamController<MotionEvent>.broadcast();

  /// Stream controller for sensor data
  final _sensorDataController = StreamController<Map<String, dynamic>>.broadcast();

  /// Stream controller for connection status
  final _connectionStatusController = StreamController<bool>.broadcast();

  /// Stream of motion events
  Stream<MotionEvent> get motionEvents => _motionEventsController.stream;

  /// Stream of sensor data
  Stream<Map<String, dynamic>> get sensorData => _sensorDataController.stream;

  /// Stream of connection status
  Stream<bool> get connectionStatus => _connectionStatusController.stream;

  /// Whether the WebSocket is connected
  bool _isConnected = false;

  /// Gets whether the WebSocket is connected
  bool get isConnected => _isConnected;

  /// The WebSocket server URL
  final String _serverUrl;

  /// Timer for reconnection attempts
  Timer? _reconnectTimer;

  /// Creates a new WebSocket service
  WebSocketService({required String serverUrl}) : _serverUrl = serverUrl {
    // Connect to the WebSocket server
    connect();
  }

  /// Connects to the WebSocket server
  void connect() {
    try {
      if (kDebugMode) {
        print('Attempting to connect to WebSocket server: $_serverUrl');
      }

      // Create a new WebSocket channel
      _channel = IOWebSocketChannel.connect(
        Uri.parse(_serverUrl),
        pingInterval: const Duration(seconds: 30), // Send ping every 30 seconds to keep connection alive
      );

      if (kDebugMode) {
        print('WebSocket channel created, setting up listeners');
      }

      // Listen for messages
      _channel!.stream.listen(
        (dynamic message) {
          _handleMessage(message);
        },
        onDone: () {
          if (kDebugMode) {
            print('WebSocket connection closed');
          }
          _handleDisconnection();
        },
        onError: (error) {
          if (kDebugMode) {
            print('WebSocket error: $error');
          }
          _handleDisconnection();
        },
      );

      // Update connection status
      _isConnected = true;
      _connectionStatusController.add(true);

      if (kDebugMode) {
        print('Connected to WebSocket server: $_serverUrl');
      }
    } catch (e) {
      if (kDebugMode) {
        print('Failed to connect to WebSocket server: $e');
        print('Stack trace: ${StackTrace.current}');
      }
      _handleDisconnection();
    }
  }

  /// Handles a message from the WebSocket server
  void _handleMessage(dynamic message) {
    try {
      if (kDebugMode) {
        print('Received WebSocket message: $message');
      }

      // Parse the message
      final data = jsonDecode(message as String) as Map<String, dynamic>;

      // Check the message type
      final type = data['type'] as String?;

      if (type == 'motion_event') {
        try {
          // Create a motion event from the data
          final event = _createMotionEventFromData(data);

          // Add the event to the stream
          _motionEventsController.add(event);

          if (kDebugMode) {
            print('Processed motion event: ${event.id} at ${event.formattedTime}');
            print('Temperature: ${event.temperature}°C, Humidity: ${event.humidity}%');
          }
        } catch (e) {
          if (kDebugMode) {
            print('Error processing motion event: $e');
            print('Event data: $data');
          }
        }
      } else if (type == 'sensor_data') {
        // Add the sensor data to the stream
        _sensorDataController.add(data);

        if (kDebugMode) {
          print('Received sensor data: Temperature: ${data['temperature']}°C, Humidity: ${data['humidity']}%');
        }
      } else if (type == 'motion_event_received' || type == 'sensor_data_received') {
        // Acknowledgment from server, just log it
        if (kDebugMode) {
          print('Received acknowledgment: ${data['message']}');
        }
      } else {
        if (kDebugMode) {
          print('Received unknown message type: $type');
          print('Message data: $data');
        }
      }
    } catch (e) {
      if (kDebugMode) {
        print('Error handling WebSocket message: $e');
        print('Raw message: $message');
        print('Stack trace: ${StackTrace.current}');
      }
    }
  }

  /// Creates a motion event from WebSocket data
  MotionEvent _createMotionEventFromData(Map<String, dynamic> data) {
    // Parse the timestamp
    DateTime timestamp;

    // Check if the timestamp is in the format "DEVICE_UPTIME:seconds"
    final timestampStr = data['timestamp'] as String;
    if (timestampStr.startsWith('DEVICE_UPTIME:')) {
      // Use current time for device uptime timestamps
      timestamp = DateTime.now();
    } else {
      // Try to parse as ISO 8601 timestamp
      timestamp = DateTime.tryParse(timestampStr) ?? DateTime.now();
    }

    // Create a unique ID if none is provided
    final id = data['id'] as String? ?? 'event_${DateTime.now().millisecondsSinceEpoch}';

    // Create the motion event
    return MotionEvent(
      id: id,
      timestamp: timestamp,
      deviceId: data['device_id'] as String,
      temperature: (data['temperature'] as num).toDouble(),
      humidity: (data['humidity'] as num).toDouble(),
      imageUrl: data['image_url'] as String? ?? 'https://via.placeholder.com/150',
    );
  }

  /// Handles a disconnection from the WebSocket server
  void _handleDisconnection() {
    // Update connection status
    _isConnected = false;
    _connectionStatusController.add(false);

    if (kDebugMode) {
      print('Disconnected from WebSocket server');
    }

    // Schedule a reconnection attempt
    _scheduleReconnect();
  }

  /// Schedules a reconnection attempt
  void _scheduleReconnect() {
    // Cancel any existing timer
    _reconnectTimer?.cancel();

    // Schedule a new reconnection attempt
    _reconnectTimer = Timer(const Duration(seconds: 5), () {
      if (kDebugMode) {
        print('Attempting to reconnect to WebSocket server...');
      }
      connect();
    });
  }

  /// Disconnects from the WebSocket server
  void disconnect() {
    // Cancel any reconnection attempts
    _reconnectTimer?.cancel();

    // Close the WebSocket channel
    _channel?.sink.close();

    // Update connection status
    _isConnected = false;
    _connectionStatusController.add(false);
  }

  /// Disposes of the WebSocket service
  void dispose() {
    // Disconnect from the WebSocket server
    disconnect();

    // Close the stream controllers
    _motionEventsController.close();
    _sensorDataController.close();
    _connectionStatusController.close();
  }
}

File: web_app/motion_detector/src/services/WebSocketService.js
/**
 * Service for handling WebSocket connections to the backend
 */
import { AppConfig } from '../utils/config';
import MotionEvent from '../models/MotionEvent';

class WebSocketService {
  /**
   * Creates a new WebSocket service
   * @param {string} serverUrl - The WebSocket server URL
   * @param {ApiService} apiService - The API service for getting the token
   */
  constructor(serverUrl = AppConfig.wsBaseUrl, apiService = null) {
    this.serverUrl = serverUrl;
    this.apiService = apiService;
    this.socket = null;
    this.isConnected = false;
    this.reconnectTimer = null;
    this.reconnectInterval = 5000; // 5 seconds

    // Event listeners
    this.motionEventListeners = [];
    this.sensorDataListeners = [];
    this.connectionStatusListeners = [];

    // Connect to the WebSocket server
    this.connect();
  }

  /**
   * Connects to the WebSocket server
   */
  connect() {
    try {
      // Always use the default token for WebSocket connections
      let url = this.serverUrl;
      let token = AppConfig.defaultToken;

      console.debug('Using default token from AppConfig for WebSocket connection');

      // Add token to URL - this is how the Django backend expects it for WebSocket connections
      // This matches the approach used in the Flutter app's WebSocketService
      if (url.includes('?')) {
        url = `${url}&token=${token}`;
      } else {
        url = `${url}?token=${token}`;
      }
      console.debug('Using token for WebSocket connection');

      console.debug(`Attempting to connect to WebSocket server: ${url}`);

      // Create a new WebSocket connection
      this.socket = new WebSocket(url);

      // Set up event handlers
      this.socket.onopen = this.handleOpen.bind(this);
      this.socket.onmessage = this.handleMessage.bind(this);
      this.socket.onclose = this.handleClose.bind(this);
      this.socket.onerror = this.handleError.bind(this);

      console.debug('WebSocket connection created, setting up listeners');
    } catch (error) {
      console.error('Failed to connect to WebSocket server:', error);
      console.error('Stack trace:', error.stack);
      this.handleDisconnection();
    }
  }

  /**
   * Handles WebSocket connection open
   * @param {Event} event - The open event
   */
  handleOpen(event) {
    console.debug('Connected to WebSocket server');
    this.isConnected = true;
    this.notifyConnectionStatus(true);
  }

  /**
   * Handles WebSocket messages
   * @param {MessageEvent} event - The message event
   */
  handleMessage(event) {
    try {
      console.debug('Received WebSocket message:', event.data);

      // Parse the message
      const data = JSON.parse(event.data);

      // Check the message type
      const type = data.type;

      if (type === 'motion_event') {
        try {
          // Create a motion event from the data
          const event = this.createMotionEventFromData(data);

          // Notify listeners
          this.notifyMotionEventListeners(event);

          console.debug(`Processed motion event: ${event.id} at ${event.formattedTime}`);
          console.debug(`Temperature: ${event.temperature.toFixed(1)}°C, Humidity: ${event.humidity.toFixed(1)}%`);
        } catch (error) {
          console.error('Error processing motion event:', error);
          console.error('Event data:', data);
        }
      } else if (type === 'sensor_data') {
        // Notify listeners
        this.notifySensorDataListeners(data);

        console.debug(`Received sensor data: Temperature: ${data.temperature}°C, Humidity: ${data.humidity}%`);
      } else {
        console.debug('Received unknown message type:', type);
      }
    } catch (error) {
      console.error('Error handling WebSocket message:', error);
      console.error('Message:', event.data);
    }
  }

  /**
   * Handles WebSocket connection close
   * @param {CloseEvent} event - The close event
   */
  handleClose(event) {
    console.debug(`WebSocket connection closed with code ${event.code}`);

    // Check close code for more information
    if (event.code === 1000) {
      console.debug('Normal closure, connection closed normally');
    } else if (event.code === 1001) {
      console.debug('Going away, server or client is going away');
    } else if (event.code === 1006) {
      console.error('Abnormal closure, connection closed abnormally');
    } else if (event.code === 1008) {
      console.error('Policy violation, message violates server policy');
    } else if (event.code === 1011) {
      console.error('Internal server error');
    } else if (event.code === 4003) {
      console.error('Authentication failed, invalid token');
    }

    if (event.reason) {
      console.debug('Close reason:', event.reason);
    }

    this.handleDisconnection();
  }

  /**
   * Handles WebSocket errors
   * @param {Event} event - The error event
   */
  handleError(event) {
    console.error('WebSocket error:', event);

    // Check if the error is related to authentication
    if (event.message && event.message.includes('401')) {
      console.error('WebSocket authentication failed. Check your token.');
    } else if (event.message && event.message.includes('403')) {
      console.error('WebSocket access forbidden. Check your permissions.');
    }

    this.handleDisconnection();
  }

  /**
   * Handles disconnection from the WebSocket server
   */
  handleDisconnection() {
    // Update connection status
    this.isConnected = false;
    this.notifyConnectionStatus(false);

    console.debug('Disconnected from WebSocket server');

    // Schedule a reconnection attempt
    this.scheduleReconnect();
  }

  /**
   * Schedules a reconnection attempt
   */
  scheduleReconnect() {
    // Clear any existing timer
    if (this.reconnectTimer) {
      clearTimeout(this.reconnectTimer);
    }

    // Schedule a new reconnection attempt
    this.reconnectTimer = setTimeout(() => {
      console.debug('Attempting to reconnect to WebSocket server...');
      this.connect();
    }, this.reconnectInterval);
  }

  /**
   * Creates a motion event from WebSocket data
   * @param {Object} data - The WebSocket data
   * @returns {MotionEvent} The motion event
   */
  createMotionEventFromData(data) {
    // Parse the timestamp
    const timestamp = new Date(data.timestamp);

    // Create a unique ID if none is provided
    const id = data.id || `event_${Date.now()}`;

    // Create the motion event
    return new MotionEvent(
      id,
      timestamp,
      data.device_id,
      parseFloat(data.temperature),
      parseFloat(data.humidity),
      data.image_url || 'https://via.placeholder.com/150'
    );
  }

  /**
   * Adds a motion event listener
   * @param {Function} listener - The listener function
   */
  addMotionEventListener(listener) {
    this.motionEventListeners.push(listener);
  }

  /**
   * Removes a motion event listener
   * @param {Function} listener - The listener function
   */
  removeMotionEventListener(listener) {
    this.motionEventListeners = this.motionEventListeners.filter(l => l !== listener);
  }

  /**
   * Notifies motion event listeners
   * @param {MotionEvent} event - The motion event
   */
  notifyMotionEventListeners(event) {
    this.motionEventListeners.forEach(listener => listener(event));
  }

  /**
   * Adds a sensor data listener
   * @param {Function} listener - The listener function
   */
  addSensorDataListener(listener) {
    this.sensorDataListeners.push(listener);
  }

  /**
   * Removes a sensor data listener
   * @param {Function} listener - The listener function
   */
  removeSensorDataListener(listener) {
    this.sensorDataListeners = this.sensorDataListeners.filter(l => l !== listener);
  }

  /**
   * Notifies sensor data listeners
   * @param {Object} data - The sensor data
   */
  notifySensorDataListeners(data) {
    this.sensorDataListeners.forEach(listener => listener(data));
  }

  /**
   * Adds a connection status listener
   * @param {Function} listener - The listener function
   */
  addConnectionStatusListener(listener) {
    this.connectionStatusListeners.push(listener);
    // Immediately notify with current status
    listener(this.isConnected);
  }

  /**
   * Removes a connection status listener
   * @param {Function} listener - The listener function
   */
  removeConnectionStatusListener(listener) {
    this.connectionStatusListeners = this.connectionStatusListeners.filter(l => l !== listener);
  }

  /**
   * Notifies connection status listeners
   * @param {boolean} connected - Whether the WebSocket is connected
   */
  notifyConnectionStatus(connected) {
    this.connectionStatusListeners.forEach(listener => listener(connected));
  }

  /**
   * Closes the WebSocket connection
   */
  close() {
    if (this.socket) {
      this.socket.close();
    }

    if (this.reconnectTimer) {
      clearTimeout(this.reconnectTimer);
    }
  }
}

export default WebSocketService;
